### PR TITLE
Add rate limiting on particular views

### LINKED ISSUE #11: Add rate limiting on particular views
Rate limit attempts on the following views:

* Login
* Password reset
* Verification

Log attempts in these views in the database and limit attempts that are allowed in a configurable time span. 
Include a CLI command that can be run from a cron job to purge old records

### PR DESCRIPTION
**Pull request by mtbrock**
Fixes #11

### CODE REVIEW

File: keg_auth/libs/authenticators.py [hunk 1]
```diff
@@ -206,6 +210,61 @@ def post(self, *args, **kwargs):
...
+    def attempt_ent(self):
+        return flask.current_app.auth_manager.entity_registry.attempt_cls
+
+    def should_limit_attempts(self):
+        try:
+            return self.attempt_ent is not None
+        except RegistryError:
+            return False
+
+    def is_attempt_blocked(self, user):
```
Comment (tjlevel12) [inline keg_auth/libs/authenticators.py]: I'm not sure the logic on this is quite right, and I don't see any tests to verify it works as expected.

Shouldn't you be getting the number of attempts within the previous `get_attempt_timespan()`, and then basic the count just on that? I'm not sure what the purpose of starting at the `lockout_period` is supposed to be.

Let's say we want to lock a view down for 5 minutes if a user tries to access it more than 3 times within a 15 minute window. I'm not sure this code will allow for that to happen.

Alternatively, let's say we want to lock a view down for 2 hours if a user tries to access it more than 5 times within a 10 minute window. The user accesses the view once every 10 minutes for an hour, and then tries 20 times within the next 5 minutes. Since we're going back by the `lockout period` and then moving forward by the `attempt timespan`, I don't think those most-recent attempts would be counted.

Comment (guruofgentoo) [reply keg_auth/libs/authenticators.py]: There are some things wrong here:
- Generally speaking, I think we should only lock after a number of consecutive unsuccessful attempts within the time period.
  - Which means, don't penalize the user for logging in successfully sometime within that block of attempts. If I'm locking after three bad attempts, and I log 2 unsuccessful before the user logs in correctly, the next bad attempt shouldn't lock.
  - That will affect how generic this logic is between logins and resets, but that can be factored out for how you grab your last "milestone" event (last successful login, or framing your window of reset attempts)
- I share TJ's concerns about the edge cases here, so it's worth another look and perhaps some additional "scenario" tests for some of the what-ifs. FWIW, if you search RRS for `login_id_is_locked` there is an alternate solution that's just missing the attempt timespan (you'd want to pull the date math out of the query there though)

File: keg_auth/cli.py [hunk 2]
```diff
@@ -38,4 +41,31 @@ def _create_user(as_superuser, no_mail, **kwargs):
...
+
+        user_ent = auth_manager.entity_registry.user_cls
+        try:
+            user = user_ent.get_by(username=username) or user_ent.get_by(email=username)
+        except InvalidRequestError:
+            user = None
+
+        assert user is not None, 'No user found with username "{}"'.format(username)
+
+        attempt_query = attempt_ent.query.filter_by(user_id=user.id)
```
Comment (guruofgentoo) [inline keg_auth/cli.py]: Once we have a user validated and in scope, I'd want to see this part of it moved to the Attempt model. I.e. I'd like to have this method available for use elsewhere in an application.

File: keg_auth/model/__init__.py [hunk 3]
```diff
@@ -461,6 +466,31 @@ def _query_permission_mapping(cls):
         return sa.union(direct, via_bundle)
 
 
+class AttemptMixin(object):
+    users_tablename = 'users'
+
+    @declared_attr
```
Comment (guruofgentoo) [inline keg_auth/model/__init__.py:472]: We don't want to link directly to a user here. Rate-limiting mitigates against a certain vector of attacks, e.g. dictionary attacks on the username or password.

In the case of a user entering a username that doesn't exist, it appears we are a) not logging the attempt, and b) no rate-limiting is happening. What we really need to be doing is:
- use a string identifier field for what the user put in on the rate-limited form
- optionally have a user_id for ease of later reporting, but make it nullable
- log all attempts, whether the input was matched or not

File: keg_auth/libs/authenticators.py [hunk 4]
```diff
@@ -296,19 +431,88 @@ def on_form_valid(self, form):
         try:
             user = self.parent.verify_user(login_id=form.login_id.data, password=form.password.data)
 
+            if self.should_limit_attempts():
```
Comment (guruofgentoo) [inline keg_auth/libs/authenticators.py:434]: If attempts are being blocked, we want to block it before the call to `verify_user`. When evaluating for blockage, we shouldn't care if the input is valid or not - just whether the request can proceed, and log the attempt regardless.

File: keg_auth/cli.py [hunk 5]
```diff
@@ -50,4 +53,25 @@ def _create_user(as_superuser, no_mail, **kwargs):
...
+        except RegistryError:
+            click.echo('No attempt class has been registered.')
+            return
+
+        user_ent = auth_manager.entity_registry.user_cls
+        user = user_ent.get_by(**{get_username_key(user_ent): username})
+
+        assert user is not None, 'No user found with username "{}"'.format(username)
+
+        attempt_query = attempt_ent.get_attempts_for_user_query(user, attempt_type=attempt_type)
```
Comment (guruofgentoo) [inline keg_auth/cli.py:71]: Put a `purge_attempts` method on the model mixin, and call it from here.

File: keg_auth/model/__init__.py [hunk 6]
```diff
@@ -489,6 +497,40 @@ def _query_permission_mapping(cls):
         return sa.union(direct, via_bundle)
 
 
+class AttemptMixin(object):
+    # Form input data, e.g. username
+    user_input = sa.Column(sa.Unicode(512), nullable=False)
```
Comment (mtbrock) [inline keg_auth/model/__init__.py:502]: After removing the link to user table, I decided not to include a `user_id` field on the attempt model. The main reason is to avoid making assumptions about the type of `user_id`. For example, some project may use UUIDs, in which case the `user_id` field in the attempt model would need to be overridden. Since it's not needed and for the sake of simplicity I just left it out.
